README.md
# Azure_Python



Getting Started with Azure SDK in Python
=========================================


Step1:  If you don't already have it, install Python.
======
        This Azure SDK is compatible with Python 2.7, 3.4, 3.5, 3.6 and 3.7.
        
sudo apt update -y && apt upgrade -y

sudo apt install python3-pip

Step2:  Install and initialize the virtual environment with the "venv" module on Python3.
======

apt-get install python3-venv

python3 -m venv .venv

cd .venv

source bin/activate

Step3:  Clone the repository
======

git clone https://github.com/KUMAR-REDDY-BAVANASI/python-azure-snippets.git


Step4:  Install the dependencies using pip.
======

cd Azure_Python/

sudo pip3 install -r requirements-azure-python.txt


Step5: For any Error came related to this, you will install these two dependencies.
======

pip3 install setuptools_rust

pip3 install --upgrade pip

sudo pip3 install -r requirements-azure-python.txt

Step6:  Fill in and export these environment variables into your current shell.
======

export AZURE_TENANT_ID={your tenant id}

export AZURE_CLIENT_ID={your client id}

export AZURE_CLIENT_SECRET={your client secret}

export AZURE_SUBSCRIPTION_ID={your subscription id}

Like,

export AZURE_TENANT_ID=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX

export AZURE_CLIENT_ID=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX

export AZURE_CLIENT_SECRET=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX

export SUBSCRIPTION_ID=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX



Step7:  Run the python script.
======

python3 filename.py


History
    1  ls
    2  git clone https://github.com/KUMAR-BAVANASI/Working_Python_Ansible_Azure.git
    3  ls
    4  cd Working_Python_Ansible_Azure/
    5  ls
    6  sudo apt update -y && apt upgrade -y
    7  sudo apt install python3-pip
    8  python3 --version
    9  pip3
   10  pip3 list
   11  mkdir .pip
   12  rm -fr .pip/
   13  cd
   14  pip3 list
   15  mkdir .pip
   16  cd .pip/
   17  vi pip.conf
   18  cd
   19  pip3 list
   20  apt-get install python3-venv
   21  ls
   22  cd Working_Python_Ansible_Azure/
   23  ls
   24  python3 -m venv .venv
   25  source .venv/bin/activate
   26  ls
   27  pip3 list
   28  pip3 install --upgrade pip
   29  pip3 list
   30  ls
   31  pip3 install -r requirements-azure-python.txt
   32  pip3 list
   33  vi requirements-azure-python.txt
   34  pip3 install -r requirements-azure-python.txt
   35  ls
   36  cd Azure_VM_DB_Python/
   37  ls
   38  cat azure_rg.py
   39  cat .env
   40  vi .env
   41  python3 azure_rg.py


Azure_VM_DB_Python/provision_vm.py
# Import the needed credential and management objects from the libraries.
from azure.common.credentials import ServicePrincipalCredentials
from azure.mgmt.resource import ResourceManagementClient
from azure.mgmt.network import NetworkManagementClient
from azure.mgmt.network.v2017_10_01.models import NetworkSecurityGroup
from azure.mgmt.network.v2017_10_01.models import SecurityRule
from azure.mgmt.compute import ComputeManagementClient
from decouple import config
import json

print(f"Provisioning a virtual machine...some operations might take a minute or two.")

# Configure Credentials
client_id = config('AZURE_CLIENT_ID')
client_secret = config('AZURE_CLIENT_SECRET')
tenant = config('AZURE_TENANT_ID')

credentials = ServicePrincipalCredentials(
  client_id = client_id,
  secret = client_secret,
  tenant = tenant
)

# Retrieve subscription ID from environment variable.
subscription_id = config('SUBSCRIPTION_ID')


# Step 1: Provision a resource group

# Obtain the management object for resources, using the credentials from the CLI login.
resource_client = ResourceManagementClient(credentials, subscription_id)

# Constants we need in multiple places: the resource group name and the region
# in which we provision resources. You can change these values however you want.
RESOURCE_GROUP_NAME = "PythonAzureExample-VM-rg"
LOCATION = "eastus"

# Provision the resource group.
rg_result = resource_client.resource_groups.create_or_update(RESOURCE_GROUP_NAME,
    {
        "location": LOCATION
    }
)


print(f"Provisioned resource group {rg_result.name} in the {rg_result.location} region")

# For details on the previous code, see Example: Provision a resource group
# at https://docs.microsoft.com/azure/developer/python/azure-sdk-example-resource-group


# Step 2: provision a virtual network

# A virtual machine requires a network interface client (NIC). A NIC requires
# a virtual network and subnet along with an IP address. Therefore we must provision
# these downstream components first, then provision the NIC, after which we
# can provision the VM.

# Network and IP address names
VNET_NAME = "python-example-vnet"
SUBNET_NAME = "python-example-subnet"
IP_NAME = "python-example-ip"
NSG_NAME = 'python-nsg'
IP_CONFIG_NAME = "python-example-ip-config"
NIC_NAME = "python-example-nic"

# Obtain the management object for networks
network_client = NetworkManagementClient(credentials, subscription_id)

# Provision the virtual network and wait for completion
poller = network_client.virtual_networks.create_or_update(RESOURCE_GROUP_NAME,
    VNET_NAME,
    {
        "location": LOCATION,
        "address_space": {
            "address_prefixes": ["10.0.0.0/16"]
        }
    }
)

vnet_result = poller.result()

print(f"Provisioned virtual network {vnet_result.name} with address prefixes {vnet_result.address_space.address_prefixes}")

# Step 3: Provision the subnet and wait for completion
poller = network_client.subnets.create_or_update(RESOURCE_GROUP_NAME, 
    VNET_NAME, SUBNET_NAME,
    { "address_prefix": "10.0.0.0/24" }
)
subnet_result = poller.result()

print(f"Provisioned virtual subnet {subnet_result.name} with address prefix {subnet_result.address_prefix}")

# Step 4: Provision an IP address and wait for completion
poller = network_client.public_ip_addresses.create_or_update(RESOURCE_GROUP_NAME,
    IP_NAME,
    {
        "location": LOCATION,
        "sku": { "name": "Standard" },
        "public_ip_allocation_method": "Static",
        "public_ip_address_version" : "IPV4"
    }
)

ip_address_result = poller.result()

print(f"Provisioned public IP address {ip_address_result.name} with address {ip_address_result.ip_address}")

# Step 5: Provision the network security group
nsg_rules = network_client.network_security_groups.create_or_update(
        RESOURCE_GROUP_NAME,
        NSG_NAME, 
        NetworkSecurityGroup(
            location = "eastus",
            security_rules = [SecurityRule(
                name = "SSH From Home",
                description = "Rule01",
                source_address_prefix='*',
                destination_address_prefix='*',
                protocol = "Tcp",
                source_port_range="*",
                destination_port_range="22",
                access='Allow',
                direction='Inbound',
                priority = 1001
            )
        ]
    )
)
nsg_result = nsg_rules.result()

print(f"Provisioned network security group {nsg_result.name}")

# Step 6: Provision the network interface client
poller = network_client.network_interfaces.create_or_update(RESOURCE_GROUP_NAME,
    NIC_NAME, 
    {
        "location": LOCATION,
        "network_security_group": { "id": nsg_result.id },
        "ip_configurations": [ {
            "name": IP_CONFIG_NAME,
            "subnet": { "id": subnet_result.id },
            "public_ip_address": {"id": ip_address_result.id }
        }]
    }
)

nic_result = poller.result()

print(f"Provisioned network interface client {nic_result.name}")

# Step 7: Provision the virtual machine

# Obtain the management object for virtual machines
compute_client = ComputeManagementClient(credentials, subscription_id)

VM_NAME = "ExampleVM"
USERNAME = "azureuser"
PASSWORD = "AzureLogin@1234"

print(f"Provisioning virtual machine {VM_NAME}; this operation might take a few minutes.")

# Provision the VM specifying only minimal arguments, which defaults to an Ubuntu 18.04 VM
# on a Standard DS1 v2 plan with a public IP address and a default virtual network/subnet.

poller = compute_client.virtual_machines.create_or_update(RESOURCE_GROUP_NAME, VM_NAME,
    {
        "location": LOCATION,
        "storage_profile": {
            "image_reference": {
                "publisher": 'Canonical',
                "offer": "UbuntuServer",
                "sku": "18.04-LTS",
                "version": "latest"
            }
        },
        "hardware_profile": {
            "vm_size": "Standard_DS1_v2"
        },
        "os_profile": {
            "computer_name": VM_NAME,
            "admin_username": USERNAME,
            "admin_password": PASSWORD
        },
        "network_profile": {
            "network_interfaces": [{
                "id": nic_result.id,
            }]
        }        
    }
)

vm_result = poller.result()
json_res = json.dumps(vm_result.as_dict())
print(json_res)

print(f"Provisioned virtual machine {vm_result.name}")